modules/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/FlowNodeActivityBehavior.java

```diff
@@ -52,5 +53,20 @@ public void signal(ActivityExecution execution, String signalName, Object signal
     // concrete activity behaviours that do accept signals should override this method;
     throw new ActivitiException("this activity doesn't accept signals");
   }
+
+  /**
+   * Register fired signals to handle race conditions within current transaction scope
+   */
+  protected void registerFiredSignalEvent(ActivityExecution execution, String eventName) {
```

Comment:
If the command is executed  in a nested state, it may result in unintended movement. The signal of the process instance processing within the parent scope command may also execute other process instance's signal subscriptions within the child scope command.

Reply:
## Test case

https://files.gitter.im/Activiti/Activiti7/s2fL/SignalEventTest.java
https://files.gitter.im/Activiti/Activiti7/b8EB/SignalEventTests.nestCommandParent.bpmn20.xml
https://files.gitter.im/Activiti/Activiti7/4Ujj/SignalEventTests.nestCommandSub.bpmn20.xml

Reply:
Signal names should be unique

Reply:
No, if the scope of the signal is a process instance, signal catch events should not be handled cross different process instances, even if the signal names are the same.

Reply:
@igdianov what are your thoughts?

Reply:
@igdianov @balsarori 
I made the pull request https://github.com/Activiti/Activiti/pull/1699 for this problem.

Reply:
@daisuke-yoshimoto
how about just adding extra info to the attribute to identify it from any other nested commands? for example adding the process instance id
```Context.getCommandContext().addAttribute(eventName + ":" + execution.getProcessInstanceId())```

Reply:
@balsarori @daisuke-yoshimoto I had reviewed #1699 by @daisuke-yoshimoto. I think it is a generally good idea to register signals and then match event subscribers by signal scope definition in the command context to avoid corner cases such as #1699.